### PR TITLE
Enable passthrough mode for read/write operations

### PR DESCRIPTION
**Pull request by amir73il**
Add support for filesystem passthrough read/write of files.

When the FUSE_PASSTHROUGH capability is enabled, the FUSE server may decide, while handling the "open" or "create" requests, if the given file can be accessed by that process in "passthrough" mode, meaning that all the further read and write operations would be forwarded by the kernel directly to the backing file rather than to the FUSE server. All requests other than read or write are still handled by the server.

This allows for an improved performance on reads and writes, especially in the case of reads at random offsets, for which no (readahead) caching mechanism would help, reducing the performance gap between FUSE and native filesystem access.

Extend also the passthrough_hp example with the new passthrough feature. This example opens a kernel backing file per FUSE inode on the first FUSE file open of that inode and closes the backing file on the release of the last FUSE file on that inode.

All opens of the same inode passthrough to the same backing file. A combination of fi->direct_io and fi->passthrough is allowed. It means that read/write operations go directly to the server, but mmap is done on the backing file.

This allows to open some fds of the inode in passthrough mode and some fd of the same inode in direct_io/passthrough_mmap mode.

### CODE REVIEW

File: include/fuse_common.h
Comment (nanonyme) [inline include/fuse_common.h:465]: This conflicts with above FUSE_CAP_DIRECT_IO_ALLOW_MMAP

Comment (nanonyme) [reply include/fuse_common.h:465]: I suppose this should simply be `1 << 29`

Comment (amir73il) [reply include/fuse_common.h]: yap. fixed. Thanks!

File: include/fuse_lowlevel.h
Comment (nanonyme) [inline include/fuse_lowlevel.h:1391]: Probably want to fix this typo resulting in static analysis failure while waiting for actual maintainers to review.

Comment (amir73il) [reply include/fuse_lowlevel.h]: done

File: lib/fuse_versionscript
Comment (bsbernd) [inline lib/fuse_versionscript:171]: That is not right here and will cause issues for distributions. Could you please put it under FUSE_3.17? We might need to change it again to 3.18 if 3.17 gets released before (it probably will because of the ABI issue) - hmm, maybe wait a few days, I hope I get a bit time tomorrow morning to look into that.

File: example/passthrough_hp.cc
Comment (bsbernd) [inline example/passthrough_hp.cc:201]: Hmm, no option for it? I personally use passthrough_hp to benchmarks for the kernel/userspace interface performance and also use it for xfstests - I think we need an option to at least disable passthrough.

Comment (bsbernd) [reply example/passthrough_hp.cc:201]: Ah sorry, ignore me, it gets late and I had missed the "fs.passthrough" condition.

File: lib/fuse_lowlevel.c
Comment (bsbernd) [inline lib/fuse_lowlevel.c:2194]: Could you please add a comment here why it is +1?

File: lib/fuse_versionscript
Comment (amir73il) [reply lib/fuse_versionscript:171]: Should I use FUSE_3.17 { } FUSE_3.4;?
What does FUSE_3.4 mean in this context?

File: lib/fuse_lowlevel.c
Comment (amir73il) [reply lib/fuse_lowlevel.c:2198]: done

File: include/fuse_common.h
Comment (trapexit) [inline include/fuse_common.h:620]: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/fuse/inode.c?h=v6.9-rc2#n1326

If I'm reading this correctly... that is not correct. It isn't if max > 1 but max > 2. The kernel check is `arg->max_stack_depth <= FILESYSTEM_MAX_STACK_DEPTH`

Comment (nanonyme) [reply include/fuse_common.h:620]: Good point. It's important to get this exactly right if you are eg running FUSE inside Docker container so you are already on top of overlayfs. The limit of 2 is *really* low.

File: lib/fuse_versionscript
Comment (bsbernd) [reply lib/fuse_versionscript:171]: https://ftp.gnu.org/old-gnu/Manuals/ld-2.9.1/html_mono/ld.html#SEC25

To my understanding "FUSE_3.4" is what the new version depends on. Given that we are at ABI version 3.12 it should be `FUSE_3.17 { } FUSE_3.12;`, afaik.

With the possibility that 3.17 needs to be replaced by 3.18 later this week.

File: include/fuse_common.h
Comment (bsbernd) [reply include/fuse_common.h:620]: I guess it relates to `In current kernel, FILESYSTEM_MAX_STACK_DEPTH is 2, which includes
	 * the FUSE layer, so max_backing_stack_depth > 1 is not supported.`

Comment (trapexit) [reply include/fuse_common.h:620]: But the literal check in FUSE is <= 2. Unless there is an additional check somewhere being done being referenced. In which case why would FUSE be checking?

Comment (nanonyme) [reply include/fuse_common.h:620]: I wonder if it would make sense to email the person who implemented the kernel feature and kindly request them to review the libfuse counterpart.

Comment (trapexit) [reply include/fuse_common.h:620]: I think I'd prefer a hybrid approach given that making it that binary choice would be awkward to undo down the road if it did change. Maybe just expand the docs to include that information? Described not too differently from what you said here.

Comment (bsbernd) [reply include/fuse_common.h:620]: @trapexit Sorry for my late reply, you might have found it already, if not, see in `do_init` 
```
if (se->conn.want & FUSE_CAP_PASSTHROUGH) {
		outargflags |= FUSE_PASSTHROUGH;
		/*
		 * outarg.max_stack_depth includes the fuse stack layer,
		 * so it is one more than max_backing_stack_depth.
		 */
		outarg.max_stack_depth = se->conn.max_backing_stack_depth + 1;```

File: lib/fuse_versionscript
Comment (amir73il) [reply lib/fuse_versionscript:171]: ok, it is a mistake that the last section is FUSE_3.12 { } FUSE_3.4 ? 
should it not be FUSE_3.12 {} FUSE_3.7 ?

File: include/fuse_common.h
Comment (amir73il) [reply include/fuse_common.h]: ok, added FUSE_BACKING_STACKED_{OVER,UNDER} macros (not enum) and explained them in documentation

File: lib/fuse_versionscript
Comment (bsbernd) [reply lib/fuse_versionscript:171]: Ouch, I didn't see that, yeah, I had used the wrong dependency value. I'm not sure if I should change it now. Maybe all is fine, as the fuse_session_loop dependency symbols are defined in 3.2.

**Comment by trapexit** [comment]
Passthrough is only allowed if `capable(CAP_SYS_ADMIN)`. In my fuse server I need to seteuid to the client's uid so no CAP_SYS_ADMIN. Setting KEEP_CAPS wouldn't be appropriate so I'm guessing the best strategy would be to seteuid back to root before calling ioctl?

**Comment by amir73il** [comment]
> Passthrough is only allowed if `capable(CAP_SYS_ADMIN)`. In my fuse server I need to seteuid to the client's uid so no CAP_SYS_ADMIN. Setting KEEP_CAPS wouldn't be appropriate so I'm guessing the best strategy would be to seteuid back to root before calling ioctl?

That's an option. Or seteuid only for specific operations that need it like creating files. 
Do you need seteuid for all of operations? Considering that permissions are already enforced also on fuse later? 

I don't remember if you told me about your filesystem but it always helps to know the big picture of you want to best answer

**Comment by amir73il** [comment]
> 
> Does this mean real CAP_SYS_ADMIN? So will not be possible under unprivileged user namespaces?

That is correct.
As the comment in the kernel code says: /* TODO: relax CAP_SYS_ADMIN once backing files are visible to lsof */
This TODO is not high on my priority, so if anyone has interest to push this forward, I encourage them to try to take on this work.

**Comment by nanonyme** [comment]
> > Does this mean real CAP_SYS_ADMIN? So will not be possible under unprivileged user namespaces?
> 
> That is correct. As the comment in the kernel code says: /* TODO: relax CAP_SYS_ADMIN once backing files are visible to lsof */ This TODO is not high on my priority, so if anyone has interest to push this forward, I encourage them to try to take on this work.

So what exactly happens when filesystem requests passthrough but isn't running privileged? Will there be a silent fallback to no passthrough?

**Comment by trapexit** [comment]
> So what exactly happens when filesystem requests passthrough but isn't running privileged? Will there be a silent fallback to no passthrough?

The ioctl call is denied and the server has to determine what it will do. Passthrough is not a transparent feature. The server must support it explicitly.

**Comment by trapexit** [comment]
Not sure what you mean. The server can do whatever it wants.

File: lib/fuse_versionscript
Comment (amir73il) [reply lib/fuse_versionscript]: @bsbernd I see the libfuse version was increased to 3.17.0.
Did you mean that the passthrough feature should increment the libfuse version to 3.18.0?

Comment (bsbernd) [reply lib/fuse_versionscript]: It is not finished yet, the version is was just increased to prepare the branch already.
I have created a separate patch/PR to add in the libfuse version an application was compiled against (https://github.com/libfuse/libfuse/pull/942), should be complete now, but I didn't have time for manual testing yet. I hope this weekend.  After that will add in auto-correct of the ABI issues for < 3.17.0 and then we should be good to release.

File: lib/fuse_versionscript
Comment (bsbernd) [inline lib/fuse_versionscript:190]: 3.17

**Comment by bsbernd** [comment]
@amir73il needs to be rebasd against master. I guess you are busy today and this week. Shall I force push into your branch in (my) evening?

**Review by bsbernd — APPROVED** [review]
Thanks Amir, LGTM.

**Comment by mxmlnkn** [comment]
I was looking through this PR and the [kernel patches](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=6ce8b2ce0d7e3a621cdc9eb66d74436ca7d0e66e), especially the definition of `struct fuse_backing_map`. If I may ask, do I understand this correctly that pass-through only works on file descriptor granularity? What I would have desired was passthrough to a byte-range pointing inside an existing file. Such finer granularities were mentioned in this [presentation](https://youtu.be/3IjdxsFPoHE?t=97), but aside from mentioning it, it was not discussed again there.

**Comment by amir73il** [comment]
> I was looking through this PR and the [kernel patches](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=6ce8b2ce0d7e3a621cdc9eb66d74436ca7d0e66e), especially the definition of `struct fuse_backing_map`. If I may ask, do I understand this correctly that pass-through only works on file descriptor granularity? What I would have desired was passthrough to a byte-range pointing inside an existing file. Such finer granularities were mentioned in this [presentation](https://youtu.be/3IjdxsFPoHE?t=97), but aside from mentioning it, it was not discussed again there.

setup passthrough to byte range was not implemented.
the API could be extended to support it in the future.
I currently have no estimation about how the implementation effort and design concerns.
I also have no use case for this functionality, so am not likely to be the one driving it upstream.

**Comment by mxmlnkn** [comment]
> > I was looking through this PR and the [kernel patches](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=6ce8b2ce0d7e3a621cdc9eb66d74436ca7d0e66e), especially the definition of `struct fuse_backing_map`. If I may ask, do I understand this correctly that pass-through only works on file descriptor granularity? What I would have desired was passthrough to a byte-range pointing inside an existing file. Such finer granularities were mentioned in this [presentation](https://youtu.be/3IjdxsFPoHE?t=97), but aside from mentioning it, it was not discussed again there.
> 
> setup passthrough to byte range was not implemented. the API could be extended to support it in the future. I currently have no estimation about how the implementation effort and design concerns. I also have no use case for this functionality, so am not likely to be the one driving it upstream.

I see, too bad. I would use it to map byte ranges inside non-compressed archives to files. It should work fine for TAR and ZIP. SquashFS would only work easily without tail-end compression but even that might be possible to get to work somehow. I don't have any experience with kernel development. Currently, I am hoping, and need to verify, that maybe the long-planned FUSE eBPF feature could be used to implement byte ranges in the future. The idea would be something like this:

 - Add parameter pre-processing via a BPF callback before each read/write call that adds a value onto the requested file offset. I.e., when reading from byte offset 0 in the FUSE file, it would map to a read from offset 3000 in the backing file descriptor.
 - These offsets, to be added to the request offsets, could be stored in a BPF hash map keyed by the inode. Each open operation would insert the byte range information into the hash map for the opened inode.

I have never worked with eBPF, so I am not sure whether this would work. Maybe modifying BPF maps is rather expensive or there are other problems. I'll need to read up on all that and try some things out.

Thanks for all your work enabling the current passthrough!

**Pull request merged**
